File: problems/leetcodecn/search-a-2d-matrix-ii/main.py
#!/usr/bin/env python3
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m, n = len(matrix), len(matrix[0])
        i, j = 0, n - 1
        while i < m and j >= 0:
            if matrix[i][j] == target:
                return True
            if matrix[i][j] < target:
                i += 1
            else:
                j -= 1
        return False


def main() -> None:
    m, n, t = map(int, input().split())
    a = [list(map(int, input().split())) for _ in range(m)]
    print(Solution().searchMatrix(a, t))


if __name__ == "__main__":
    main()

File: problems/leetcodecn/search-a-2d-matrix-ii/main.cpp
/**
 * Author by Rainboy
 */
// main.cpp：从右上角出发，小于 target 向下，大于 target 向左，O(m+n)。
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size(), n = matrix[0].size();
        int i = 0, j = n - 1;
        while (i < m && j >= 0) {
            if (matrix[i][j] == target) return true;
            if (matrix[i][j] < target) i++;
            else j--;
        }
        return false;
    }
};

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr);
    int m, n, t; cin >> m >> n >> t;
    vector<vector<int>> a(m, vector<int>(n));
    for (int i = 0; i < m; i++)
        for (int j = 0; j < n; j++)
            cin >> a[i][j];
    cout << Solution().searchMatrix(a, t) << '\n';
    return 0;
}
